runs/restore.sh

```diff
@@ -1,20 +1,51 @@
 #!/bin/bash
-echo "****************************************"
-echo "Step 1: moving file to home directory..."
-sudo cp /var/www/html/openWB/web/tools/upload/backup.tar.gz /home/pi/backup.tar.gz
-cd /home/pi/
-echo "****************************************"
-echo "Step 2: extracting archive..."
-sudo tar -vxf backup.tar.gz
-echo "****************************************"
-echo "Step 3: replacing old files..."
-sudo cp -v -R /home/pi/var/www/html/openWB/* /var/www/html/openWB/
-sudo chmod 777 /var/www/html/openWB/openwb.conf
-echo "****************************************"
-echo "Step 4: cleanup after restore..."
-sudo rm /var/www/html/openWB/web/tools/upload/backup.tar.gz
-sudo rm /home/pi/backup.tar.gz
-sudo rm -R /home/pi/var
-echo "****************************************"
-echo "End: Restore finished."
-echo "****************************************"
+OPENWBBASEDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+RAMDISKDIR="$OPENWBBASEDIR/ramdisk"
+SOURCEFILE="$OPENWBBASEDIR/web/tools/upload/backup.tar.gz"
+WORKINGDIR="/home/pi/openwb_restore"
+MOSQUITTODIR="/var/lib/mosquitto"
+LOGFILE="$OPENWBBASEDIR/web/tools/upload/restore.log"
+
+{
+	echo "$(date +"%Y-%m-%d %H:%M:%S") Restore of backup started..."
+	echo "****************************************"
+	echo "Step 1: setting flag 'update in progress' and wait for control loop to finish"
+	echo 1 >"$OPENWBBASEDIR/ramdisk/updateinprogress"
+	# Wait for regulation loop(s) and cron jobs to end, but with timeout in case a script hangs
+	pgrep -f "$OPENWBBASEDIR/(regel\\.sh|runs/cron5min\\.sh|runs/cronnightly\\.sh)$" |
+		timeout 15 xargs -n1 -I'{}' tail -f --pid="{}" /dev/null
+	echo "****************************************"
+	echo "Step 2: creating working directory \"$WORKINGDIR\""
+	mkdir -p "$WORKINGDIR"
+	echo "****************************************"
+	echo "Step 3: extracting archive to working dir \"$WORKINGDIR\"..."
+	if ! sudo tar --verbose --extract --file="$SOURCEFILE" --directory="$WORKINGDIR"; then
+		echo "something went wrong! aborting restore"
+		echo "Wiederherstellung fehlgeschlagen! Bitte Protokolldateien prüfen." >"$RAMDISKDIR/lastregelungaktiv"
+		echo 0 >"$OPENWBBASEDIR/ramdisk/updateinprogress"
+		exit 1
+	fi
+	echo "****************************************"
+	echo "Step 4: replacing old files..."
+	# we use cp not mv because of not empty directories in destination
+	# dotglob is needed to include files and directories beginning with a dot
+	(shopt -s dotglob; sudo cp -v -p -r "${WORKINGDIR}${OPENWBBASEDIR}/"* "${OPENWBBASEDIR}/")
```

Comment:
Das `shopt` ist nicht nötig. Es täte auch:
```bash
cp -v -p -r "${WORKINGDIR}${OPENWBBASEDIR}/." "${OPENWBBASEDIR}"
```
Das "Geheimnis" ist `/.` statt `/*`. Hat nebenbei den Vorteil dass es nicht darunter leidet die maximale Kommandozeilenlänge überschreiten zu können.

Damit ist es immernoch so, dass Dateien die bereits existieren zwar überschrieben werden, aber es werden keine Dateien gelöscht, die im Backup nicht enthalten sind. Ich denke immernoch dass der einfachste Weg `mv` wäre. In etwa so (nur so modellhaft aus dem Kopf genommen):
```bash
mv "$OPENWBBASEDIR" "$OPENWBBASEDIR_old"
mv "${WORKINGDIR}${OPENWBBASEDIR}" "$OPENWBBASEDIR"
rm -r "$OPENWBBASEDIR_old"
```
Das ganze hat natürlich das Risiko, dass wenn der Prozess aus irgend einem Grund zwischen dem ersten und zweiten `mv` unterbrochen wird, dann steht man dumm da. Das Risiko ist aber relativ gering, da es sich um eine recht simple und schnelle Instruktion handelt.

Eine andere Möglichkeit wäre rsync:
```bash
rsync -r --archive --ignore-times --delete "${WORKINGDIR}${OPENWBBASEDIR}/" "${OPENWBBASEDIR}"
```
(auch der Befehl ist eben schnell aus dem Kopf mit man-Page zusammengeworfen und sollte vor Verwendung überprüft werden. Aber ja bei `rsync` muss der Pfad mit `/` enden, wenn man nicht den Ordner sondern den Inhalt kopieren will).
`rsync` hat nämlich die Möglichkeit die nicht existierenden Dateien am Ziel zu löschen (danke `--delete`). Bei `rsync` besteht genau wie bei `cp` das Risiko, dass wenn der Prozess zwischendurch abbricht, ein inkonsistenter Status hinterlassen wird. Das Problem besteht bei `mv` nicht.

Der aktuelle Ansatz hat natürlich auch einen "Vorteil": Die daily&monthly-CSV-logs die seit dem Backup angelegt wurden bleiben bei einem Restore erhalten. Das ist allerdings in Anführungszeichen, denn wenn fehlerhafte CSV-Dateien gerade dazu führen, dass jemand ein Backup einspielen möchte dann ist damit nicht geholfen.

Reply:
Bist du sIcher das beim mv der Mountpint der Ramdisk nicht im _old landet.
Bis zum reboot / remount ist die Ramdisk dann erst mal ein normales Verzeichniss (oder fehlt ganz)

Reply:
Ach die Ramdisk habe ich ganz vergessen. Ja, die würde dann natürlich in _old landen. Die müsste man dann umounten, dann ohne ramdisk wiederherstellen und dann re-mounten. Und die Ramdisk (falls überhaupt) erst nach dem re-mount wiederherstellen.

Reply:
Die Änderung am cp nehme ich rein. Der Rest sprengt den Rahmen dieses PR.